File: 1671_minimum_number_of_removals_to_make_mountain_array/solution.py
class Solution:
    def minimumMountainRemovals(self, nums):
        from bisect import bisect_left
        def lis(a):
            d=[];out=[]
            for x in a:
                i=bisect_left(d,x)
                if i==len(d):d.append(x)
                else:d[i]=x
                out.append(i+1)
            return out
        l=lis(nums);r=lis(nums[::-1])[::-1];n=len(nums)
        return n-max((l[i]+r[i]-1 for i in range(n) if l[i]>1 and r[i]>1),default=0)

File: 1671_minimum_number_of_removals_to_make_mountain_array/Solution.java
// LeetCode 1671 - Minimum Number of Removals to Make Mountain Array
// https://leetcode.com/problems/minimum-number-of-removals-to-make-mountain-array/

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

class Solution {
    public int minimumMountainRemovals(int[] nums) {
        int n = nums.length;
        int[] left = lis(nums);
        int[] reversed = new int[n];
        for (int i = 0; i < n; i++) {
            reversed[i] = nums[n - 1 - i];
        }
        int[] rightRev = lis(reversed);
        int[] right = new int[n];
        for (int i = 0; i < n; i++) {
            right[i] = rightRev[n - 1 - i];
        }
        int best = 0;
        for (int i = 0; i < n; i++) {
            if (left[i] > 1 && right[i] > 1) {
                best = Math.max(best, left[i] + right[i] - 1);
            }
        }
        return n - best;
    }

    private int[] lis(int[] a) {
        List<Integer> tails = new ArrayList<>();
        int[] out = new int[a.length];
        for (int i = 0; i < a.length; i++) {
            int x = a[i];
            int idx = Collections.binarySearch(tails, x);
            if (idx < 0) {
                idx = -idx - 1;
            }
            if (idx == tails.size()) {
                tails.add(x);
            } else {
                tails.set(idx, x);
            }
            out[i] = idx + 1;
        }
        return out;
    }
}
